### PR TITLE
Issue #1622 port `du` to windows

### LINKED ISSUE #1622: Port du on Windows
Hi,
I downloaded 0.0.1.1 version via scoop but I don't see du on function list.
```
PS D:\Downloads\coreutils-0.0.1.1-x86_64-pc-windows-msvc> .\coreutils.exe
coreutils 0.0.1 (multi-call binary)

Usage: coreutils [function [arguments...]]

Currently defined functions/utilities:

    arch, base32, base64, basename, cat, cksum, comm, cp, cut, date, df, dircolors, dirname,
    echo, env, expand, expr, factor, false, fmt, fold, hashsum, head, hostname, join, link, ln,
    ls, md5sum, mkdir, mktemp, more, mv, nl, nproc, od, paste, printenv, printf, ptx, pwd,
    readlink, realpath, relpath, rm, rmdir, seq, sha1sum, sha224sum, sha256sum, sha3-224sum,
    sha3-256sum, sha3-384sum, sha3-512sum, sha384sum, sha3sum, sha512sum, shake128sum,
    shake256sum, shred, shuf, sleep, sort, split, sum, sync, tac, tail, tee, test, touch, tr,
    true, truncate, tsort, unexpand, uniq, wc, whoami, yes
```

What I'm doing wrong?

### PR DESCRIPTION
**Pull request by paulotten**
Fixes #1622

This is my attempt at porting `du` to Windows.

Inodes as far as I know don't exist on Windows, so I've omitted them.

If anyone knows of a better way of calculating ~~block size~~ number of blocks on Windows I'm happy to hear it.

Finally 4 out of 6 tests are failing, mainly due to different block sizes being reported. There already seems to be three sets of expected results (linux, macos, wsl). Possibly Windows deserves it's own.

### CODE REVIEW

**Comment by paulotten** [comment]
So I need to pass those tests. Too many checks are failing due to them.

Can someone please review where I'm calculating number of blocks? https://github.com/uutils/coreutils/pull/1788/files#diff-8c6f90e16c56b8fafaf6832528f61d573f94a2e3dbd28cd8a317c2cf7973322cR80

Does it look like I should be fixing my code or fixing the tests?

File: src/uu/du/src/du.rs [hunk 1]
```diff
@@ -61,15 +64,39 @@ impl Stat {
             path,
             is_dir: metadata.is_dir(),
             size: metadata.len(),
+
+            #[cfg(not(windows))]
```
Comment (sylvestre) [inline src/uu/du/src/du.rs:68]: this is a lot of "#[cfg(not(windows))]"
would it be possible to find a better way?

**Review by sylvestre — CHANGES_REQUESTED** [review]
Could you please fix the tests too? thanks!

File: src/uu/du/src/du.rs [hunk 2]
```diff
@@ -58,33 +58,31 @@ struct Stat {
...
             created: metadata.mtime() as u64,
-            #[cfg(not(windows))]
             accessed: metadata.atime() as u64,
-            #[cfg(not(windows))]
             modified: metadata.mtime() as u64,
-
-            #[cfg(windows)]
+        })
+    }
+    #[cfg(windows)]
```
Comment (sylvestre) [inline src/uu/du/src/du.rs:75]: isn't it possible to do better? It is a bunch of duplicated code

**Comment by sylvestre** [comment]
@paulotten I guess you noticed that the tests are failing on Windows
```

failures:

---- test_du::test_du_d_flag stdout ----
current_directory_resolved: 
run: D:\a\coreutils\coreutils\target\i686-pc-windows-gnu\debug\coreutils.exe du -d 1
thread 'test_du::test_du_d_flag' panicked at 'assertion failed: `(left == right)`
  left: `"6\t./subdir\n7\t./\n"`,
 right: `"28\t./subdir\n36\t./\n"`', D:\a\coreutils\coreutils\tests\by-util\test_du.rs:144:9

---- test_du::test_du_basics_subdir stdout ----
current_directory_resolved: 
run: D:\a\coreutils\coreutils\target\i686-pc-windows-gnu\debug\coreutils.exe du subdir/deeper
thread 'test_du::test_du_basics_subdir' panicked at 'assertion failed: `(left == right)`
  left: `"1\tsubdir/deeper\n"`,
 right: `"8\tsubdir/deeper\n"`', D:\a\coreutils\coreutils\tests\by-util\test_du.rs:52:9
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace

---- test_du::test_du_soft_link stdout ----
current_directory_resolved: 
run: ln -s subdir/links/subwords.txt subdir/links/sublink.txt
thread 'test_du::test_du_soft_link' panicked at 'assertion failed: link.success', D:\a\coreutils\coreutils\tests\by-util\test_du.rs:75:5

---- test_du::test_du_hard_link stdout ----
current_directory_resolved: 
run: ln subdir/links/subwords.txt subdir/links/sublink.txt
run: D:\a\coreutils\coreutils\target\i686-pc-windows-gnu\debug\coreutils.exe du subdir/links
thread 'test_du::test_du_hard_link' panicked at 'assertion failed: `(left == right)`
  left: `"11\tsubdir/links\n"`,
 right: `"16\tsubdir/links\n"`', D:\a\coreutils\coreutils\tests\by-util\test_du.rs:120:9


failures:
    test_du::test_du_basics_subdir
    test_du::test_du_d_flag
    test_du::test_du_hard_link
    test_du::test_du_soft_link
```

**Comment by paulotten** [comment]
> I guess you noticed that the tests are failing on Windows

Yes. I've added a winapi call to get the actual size on disk. I'm now seeing the same results as on WSL.

`ln` is still failing.

```
run: ln subdir/links/subwords.txt subdir/links/sublink.txt
thread 'test_du::test_du_hard_link' panicked at 'called `Result::unwrap()` on an `Err` value: Os { code: 2, kind: NotFound, message: "The system cannot find the file specified." }', tests\common\util.rs:562:14
```

and 

```
run: ln -s subdir/links/subwords.txt subdir/links/sublink.txt
thread 'test_du::test_du_soft_link' panicked at 'called `Result::unwrap()` on an `Err` value: Os { code: 2, kind: NotFound, message: "The system cannot find the file specified." }', tests\common\util.rs:562:14
```

Any insight on why would be appreciated. I obviously haven't changed any `ln` code with this pull request. If necessary I could disable those tests on windows.

**Comment by paulotten** [comment]
@sylvestre poke

**Comment by sylvestre** [comment]
The test failing is about du
```
---- test_du::test_du_soft_link stdout ----
current_directory_resolved: 
run: ln -s subdir/links/subwords.txt subdir/links/sublink.txt
thread 'test_du::test_du_soft_link' panicked at 'assertion failed: link.success', D:\a\coreutils\coreutils\tests\by-util\test_du.rs:79:5
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace

```

you should println stderr and stdout to help with debugging and see what is wrong

**Comment by paulotten** [comment]
As mentioned in https://github.com/uutils/coreutils/pull/1788#issuecomment-803437812 I'm seeing different errors locally.

Please review the output of `cargo test test_du -- --nocapture`

```
running 6 tests
current_directory_resolved: 
current_directory_resolved:
current_directory_resolved:
current_directory_resolved:
current_directory_resolved:
current_directory_resolved:
run: ln subdir/links/subwords.txt subdir/links/sublink.txt
run: C:\rust\coreutils\target\debug\coreutils.exe du subdir/deeper
run: ln -s subdir/links/subwords.txt subdir/links/sublink.txt
run: C:\rust\coreutils\target\debug\coreutils.exe du -d 1
run: C:\rust\coreutils\target\debug\coreutils.exe du bad_name
run: C:\rust\coreutils\target\debug\coreutils.exe du
thread 'test_du::test_du_hard_link' panicked at 'called `Result::unwrap()` on an `Err` value: Os { code: 2, kind: NotFound, message: "The system cannot find the file specified." }', tests\common\util.rs:562:14
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
test test_du::test_du_hard_link ... FAILED
thread 'test_du::test_du_soft_link' panicked at 'called `Result::unwrap()` on an `Err` value: Os { code: 2, kind: NotFound, message: "The system cannot find the file specified." }', tests\common\util.rs:562:14
test test_du::test_du_soft_link ... FAILED
test test_du::test_du_d_flag ... ok
test test_du::test_du_basics ... ok
test test_du::test_du_basics_subdir ... ok
test test_du::test_du_basics_bad_name ... ok

failures:

failures:
    test_du::test_du_hard_link
    test_du::test_du_soft_link

test result: FAILED. 4 passed; 2 failed; 0 ignored; 0 measured; 692 filtered out

error: test failed, to rerun pass '--test tests'
```

Note that the 4 passed tests match to corresponding `run: C:\rust\coreutils\target\debug\coreutils.exe du` commands. The two failed tests however match to `run: ln` commands.

The 4 passed tests appear to be using `coreutils`. The two failed however appear to be looking for `ln` on the system path. (I had previously assumed that the two failed tests were using the `coreutils` version of `ln`.) On a clean Windows install `ln` will not be found on the system path. It seems to me that these two tests, as currently written, are not valid on a Windows system.

@sylvestre thoughts?

**Comment by sylvestre** [comment]
the error is "The system cannot find the file specified"

This command fails?
```
run: ln -s subdir/links/subwords.txt subdir/links/sublink.txt
```

**Comment by paulotten** [comment]
This fails

```
run: ln subdir/links/subwords.txt subdir/links/sublink.txt
```

This also fails

```
run: ln -s subdir/links/subwords.txt subdir/links/sublink.txt
```

---

I can strip `test_du_soft_link` down to just

```
fn test_du_soft_link() {
    let ts = TestScenario::new("du");

    ts.cmd("ln").run();
}
```

and it still fails with

```
thread 'test_du::test_du_soft_link' panicked at 'called `Result::unwrap()` on an `Err` value: Os { code: 2, kind: NotFound, message: "The system cannot find the file specified." }', tests\common\util.rs:562:14
```

**Comment by paulotten** [comment]
https://github.com/uutils/coreutils#installation-instructions

`cargo install --path .` puts `coreutils` on my path. So `coreutils ln` works afterwards in command prompt or powershell. `ln` still doesn't resolve though.

**Comment by Artoria2e5** [comment]
I think it would still be more... informative to keep the tests on, but rewrite them to call `coreutils ln`. I am inclined to add a method to `crate::tests::common::util::TestScenario` so that:
```rust
    /// Returns builder for invoking any uutils command. Paths given are treated
    /// relative to the environment's unique temporary test directory.
    pub fn ccmd<S: AsRef<OsStr>>(&self, bin: S) -> UCommand {
        let mut cmd = self.cmd(&self.bin_path);
        cmd.arg(self.tmpd.clone());
        cmd
    }
```

... so you can just replace the `.cmd("ln")` with `.ccmd("ln")`.

**Comment by Artoria2e5** [comment]
And WRT inode numbers, Windows does have them for some filesystems (NTFS, ReFS, SMB), but they like to call it a "fileID". Like inodes these things uniquely identify files on filesystems with a concept of hard links, and should be somewhat important when you are trying to not double-count files.

The API to get it is GetFileInformationByHandle. Now here's the spicy thing: the `Stat` struct used by this project is only 64 bits, while Windows fileIDs are technically 128 bits (it used to be 64).

**Comment by paulotten** [comment]
> I think it would still be more... informative to keep the tests on, but rewrite them to call `coreutils ln`.

Going down this route, any idea why I get `ln: error: A required privilege is not held by the client. (os error 1314)` locally when I try `coreutils ln -s` or tests that do the same?

The downside of this approach seems to be that I'm now testing both `du` and `ln`.

**Comment by Artoria2e5** [comment]
@paulotten Ah, that! Symlinking by non-admin on Windows is only supported on Windows 10, and it needs to be enabled from some developer settings. Let me check if there's a programmatic way (read: registry) to do that for the CI.

Okay, it's apparently just `reg add "HKEY_LOCAL_MACHINE\SOFTWARE\Microsoft\Windows\CurrentVersion\AppModelUnlock" /t REG_DWO RD /f /v "AllowDevelopmentWithoutDevLicense" /d "1"`. Just run it before the test fires from the CI files.

**Comment by paulotten** [comment]
It appears to be

```
reg add "HKEY_LOCAL_MACHINE\SOFTWARE\Microsoft\Windows\CurrentVersion\AppModelUnlock" /t REG_DWORD /f /v "AllowDevelopmentWithoutDevLicense" /d "1"
```

You had an extra space midword. It also needs to be run as admin. `coreutils ln -s` works for me after running this.

Is this documented in `coreutils` anywhere? If not should I open an issue about adding it to the documentation?

**Comment by paulotten** [comment]
That should be `test_du_soft_link` working. Looking at `test_du_hard_link` next.

**Comment by Artoria2e5** [comment]
I don't think it's document anywhere. I recommend doing a documentation PR to rust-lang/rust for their windows fs symlink thing in std, so that on uutils side it can be used a bit more easily. Maybe I'll do that PR.

**Comment by paulotten** [comment]
> I recommend doing a documentation PR to rust-lang/rust for their windows fs symlink thing in std, so that on uutils side it can be used a bit more easily. Maybe I'll do that PR.

@Artoria2e5 I'm fine with you taking that.

I'll focus on finishing this issue.

**Comment by paulotten** [comment]
It looks like I'm passing all the checks now.

@sylvestre, @Artoria2e5 can I get a rereview please?

**Comment by sylvestre** [comment]
well done :)

**Pull request merged**